example/advection/advection_package.cpp

```diff
@@ -287,76 +294,110 @@ TaskStatus CalculateFluxes(Container<Real> &rc) {
   IndexRange ib = pmb->cellbounds.GetBoundsI(IndexDomain::interior);
   IndexRange jb = pmb->cellbounds.GetBoundsJ(IndexDomain::interior);
   IndexRange kb = pmb->cellbounds.GetBoundsK(IndexDomain::interior);
-  int ncellsi = pmb->cellbounds.ncellsi(IndexDomain::entire);
 
-  CellVariable<Real> &q = rc.Get("advected");
+  CellVariable<Real> &advected = rc.Get("advected");
   auto pkg = pmb->packages["advection_package"];
   const auto &vx = pkg->Param<Real>("vx");
   const auto &vy = pkg->Param<Real>("vy");
   const auto &vz = pkg->Param<Real>("vz");
 
-  ParArrayND<Real> ql("ql", q.GetDim(4), ncellsi);
-  ParArrayND<Real> qr("qr", q.GetDim(4), ncellsi);
-  ParArrayND<Real> qltemp("qltemp", q.GetDim(4), ncellsi);
-
+  const int scratch_level = 1; // 0 is actual scratch (tiny); 1 is HBM
+  const int nx1 = pmb->cellbounds.ncellsi(IndexDomain::entire);
+  const int nvar = advected.GetDim(4);
+  size_t scratch_size_in_bytes = parthenon::ScratchPad2D<Real>::shmem_size(nvar, nx1);
+  parthenon::ParArray4D<Real> x1flux = advected.flux[X1DIR].Get<4>();
   // get x-fluxes
-  for (int k = kb.s; k <= kb.e; k++) {
-    for (int j = jb.s; j <= jb.e; j++) {
-      // get reconstructed state on faces
-      pmb->precon->DonorCellX1(k, j, ib.s - 1, ib.e + 1, q.data, ql, qr);
-      if (vx > 0.0) {
-        for (int i = ib.s; i <= ib.e + 1; i++) {
-          q.flux[X1DIR](k, j, i) = ql(i) * vx;
-        }
-      } else {
-        for (int i = ib.s; i <= ib.e + 1; i++) {
-          q.flux[X1DIR](k, j, i) = qr(i) * vx;
+  pmb->par_for_outer(
+      "x1 flux", 2 * scratch_size_in_bytes, scratch_level, kb.s, kb.e, jb.s, jb.e,
+      KOKKOS_LAMBDA(parthenon::team_mbr_t member, const int k, const int j) {
+        parthenon::ScratchPad2D<Real> ql(member.team_scratch(scratch_level), nvar, nx1);
+        parthenon::ScratchPad2D<Real> qr(member.team_scratch(scratch_level), nvar, nx1);
+        // get reconstructed state on faces
+        parthenon::DonorCellX1(member, k, j, ib.s - 1, ib.e + 1, advected.data, ql, qr);
+        // Sync all threads in the team so that scratch memory is consistent
+        member.team_barrier();
+
+        for (int n = 0; n < nvar; n++) {
+          if (vx > 0.0) {
+            parthenon::par_for_inner(member, ib.s, ib.e + 1, [&](const int i) {
+              x1flux(n, k, j, i) = ql(n, i) * vx;
+            });
+          } else {
+            parthenon::par_for_inner(member, ib.s, ib.e + 1, [&](const int i) {
+              x1flux(n, k, j, i) = qr(n, i) * vx;
+            });
+          }
```

Comment:
I understand this nested parallelism design is almost certainly required to get things to vectorize on the CPU. However, I wonder what the impacts of it are on the GPU. Do you have any insight, @forrestglines @pgrete ? 

It might be worth experimenting with disabling it for some architectures.

Reply:
The main reason for nested parallelism is not necessarily vectorization (this also worked fine in K-Athena without it both on CPUs and GPUs) but the use of temporary/scratch pad arrays (the original scrach pad arrays as MeshBlock member variables cannot be used any more due to the different threading model when using GPUs).
Ideally, all architectures will benefit from this as it allows to reduce data movement and increased data locality.

Reply:
Huh. I always thought it was because the code didn't vectorize without pre-fetching. 

Is it clear that explicitly creating the scratch arrays is actually more performant on GPUs? My naive concern would be that the requirement of synchronizing these arrays may actually hurt us on high-bandwidth architectures? I don't have any justification for that concern.

Reply:
@Yurlungur On the GPUs with scratch level 0, scratch memory has a higher bandwidth than DRAM. Just going from the roofline models we explored with K-Athena, the code appeared to be limited by the bandwidth of DRAM memory. Merging some kernels to keep things in scratch memory and out of DRAM memory should let us cross the DRAM performance boundary.  We might also implement something like this: https://devblogs.nvidia.com/finite-difference-methods-cuda-cc-part-1/ which I've seen work very well in the past. 

Synchronization happens within a team on a single CUDA SM, which I haven't seen as a problem, but it might be on other architectures. The bigger issue I've seen in the past is that the register usage for the larger kernels can grow to the point that the GPU can't run as many concurrent threads and you loose performance. We'll have to rely on profiling to determine if that's the case.

I can't speak on how this will translate for AMD or Intel GPUs. My only bet is that the graphics pipelines are similar enough that the hardware will be comparable.

Reply:
Thanks for the explanation, @forrestglines ! 

I guess the take-away from this is that we need to do some profiling as we move forward. But it's interesting to hear that you already made some roofline plots for K-Athena and found the limit was the DRAM bandwidth.